File: desktop/Backend-Rust/src/models/focus_session.rs
// Focus Session models - focus tracking sessions stored in Firestore
// Path: users/{uid}/focus_sessions/{session_id}

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Focus status - whether user is focused or distracted
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum FocusStatus {
    Focused,
    Distracted,
}

impl std::fmt::Display for FocusStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            FocusStatus::Focused => write!(f, "focused"),
            FocusStatus::Distracted => write!(f, "distracted"),
        }
    }
}

/// Focus session stored in Firestore subcollection
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FocusSessionDB {
    /// Document ID
    pub id: String,
    /// Focus status: focused or distracted
    pub status: FocusStatus,
    /// The app or website visible
    pub app_or_site: String,
    /// Brief description of what's on screen
    pub description: String,
    /// Optional coaching message
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    /// When the session was created
    pub created_at: DateTime<Utc>,
    /// Duration in seconds (if known)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_seconds: Option<i64>,
}

/// Entry for a distraction app/site with time spent
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DistractionEntry {
    /// App or website name
    pub app_or_site: String,
    /// Total time spent in seconds
    pub total_seconds: i64,
    /// Number of times visited
    pub count: i64,
}

/// Focus statistics for a given day
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FocusStats {
    /// Date in YYYY-MM-DD format
    pub date: String,
    /// Total focused minutes
    pub focused_minutes: i64,
    /// Total distracted minutes
    pub distracted_minutes: i64,
    /// Number of sessions
    pub session_count: i64,
    /// Number of focused sessions
    pub focused_count: i64,
    /// Number of distracted sessions
    pub distracted_count: i64,
    /// Top distraction apps/sites
    pub top_distractions: Vec<DistractionEntry>,
}

File: backend/models/focus_session.py
"""Focus sessions — focus/distraction tracking and statistics.

Response wire shapes for /v1/focus-sessions* and /v1/focus-stats. Source of
truth for the focus-session response schema; routers/database construct dicts
matching these fields.

Collection: users/{uid}/focus_sessions.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class FocusSession(BaseModel):
    """A single focus/distraction session record."""

    id: str = Field(description='Unique session identifier.')
    status: str = Field(description='Session status, "focused" or "distracted".')
    app_or_site: str = Field(description='App or website the session refers to.')
    description: str = Field(description='Free-text description of the session.')
    message: Optional[str] = Field(default=None, description='Optional user message attached to the session.')
    created_at: datetime = Field(description='When the session was recorded (UTC).')
    duration_seconds: Optional[int] = Field(default=None, ge=0, description='Session duration in seconds, if known.')


class FocusDistraction(BaseModel):
    """One entry in the top-distractions breakdown of /v1/focus-stats."""

    app_or_site: str = Field(description='App or website that caused the distraction.')
    total_seconds: int = Field(ge=0, description='Total distracted seconds attributed to this app/site.')
    count: int = Field(ge=0, description='Number of distracted sessions for this app/site.')


class FocusStats(BaseModel):
    """Aggregated focus statistics returned by /v1/focus-stats."""

    date: str = Field(description='The date (YYYY-MM-DD) the stats cover.')
    focused_minutes: int = Field(ge=0, description='Total focused time, in whole minutes.')
    distracted_minutes: int = Field(ge=0, description='Total distracted time, in whole minutes.')
    session_count: int = Field(ge=0, description='Total number of sessions considered.')
    focused_count: int = Field(ge=0, description='Number of focused sessions.')
    distracted_count: int = Field(ge=0, description='Number of distracted sessions.')
    top_distractions: list[FocusDistraction] = Field(
        default_factory=list,
        description='Up to five most distracting apps/sites by total distracted seconds.',
    )
